packages/namekit-react/src/components/TruncatedText.tsx

```diff
@@ -0,0 +1,127 @@
+/**
+ * This TruncatedText component is a duplicate of @namehash/nameguard-react
+ * TruncatedText component. This duplicate will, in the near future, be the
+ * only TruncatedText component in the NameKit libraries. The duplicate is
+ * necessary for the moment so that it is included in main branch while
+ * @namehash/nameguard-react still has its TruncatedText: why is that?
+ *
+ * So any imports of @namehash/nameguard-react TruncatedText component
+ * do not break once this change is merged into main! Once this duplicate
+ * is in main, NameHash Labs will remove the TruncatedText component from
+ * @namehash/nameguard-react and update the imports to use this component
+ * instead, the @namehash/namekit-react TruncatedText.
+ *
+ * A ShortCut story was created to track this migration:
+ * https://app.shortcut.com/ps-web3/story/25261/next-steps-on-displayedname-and-truncatedtext
+ */
+
+import React, { useEffect, useState, useRef } from "react";
+import { Tooltip } from "./Tooltip";
+import cc from "classcat";
+
+interface TruncatedTextProps {
+  text: string;
+  maxTooltipWidth?: number;
+  maxDisplayWidth?: number;
+  textStylingClasses?: string;
+  tooltipTextStylingClasses?: string;
+  displayTooltipWhenTextOverflows?: boolean;
```

Comment:
I would consider adding two additional props:

```ts
renderTooltip,
tooltipProps = {},
```

This makes the component more composable that allows us to pass a custom tooltip structure if necessary. Think instances of Popover card vs Tooltip.

Reply:
@notrab Interesting.

Am I correct to assume that if `renderTooltip` prop is undefined (the default) then the `TruncatedText` UI component would then just use the default `Tooltip` UI component?

If so, could you please create a new PR for Franco's PR 381 that would implement your suggestions? Please also include:
1. Comments on the newly added props describing how they work / rules for how they work.
2. Storybook stories with examples.

Alternatively, if the scope for completing the above isn't sufficiently small to action right now, then can you please create a new Shortcut story for us to track these ideas and then we can come back to them another time?

Reply:
@notrab @FrancoAguzzi If we made `TruncatedText` more composable so that a custom `Tooltip` (or other component could be passed in) then would it also make sense to remove the following props from `TruncatedText`?

1. `maxTooltipWidth`
2. `tooltipTextStylingClasses`

If someone wanted to customize these then it seems they should just make use of the new `renderTooltip` prop to achieve their goals instead?

Reply:
Concisely written, @lightwalker-eth 

And great ideation for lib's value increasing, @notrab

Reply:
I am awaiting for idea alignment before implementing this 😉

Reply:
@FrancoAguzzi I'd like to see this action taken. It's very important to make these UI components composable. For me its a sign of bad design to see props like `maxTooltipWidth` or `tooltipTextStylingClasses` on `TruncatedText`. It's a sign that we failed to make our components composable.

Reply:
@lightwalker-eth happy to bootstrap this with few specs. If any adjustments are needed, let's discuss this together.

**TLDR;**
I suggest a very thin `TruncatedText` with `text`, `textStylingClasses`, `maxDisplayWidth` and `onTextOverflow` props.
I suggest a new component called `TextWithTooltipForOverflows` that integrates `TruncatedText` and `Tooltip`.
I suggest we export `TruncatedText` with the new `onTextOverflow` prop so integrators get to know when to display their custom tooltips and do this within their code.
Why is that? Triggering an _external_ tooltip and managing its props in relation to our `TruncatedText` props can be way more complex! Tooltip positioning will probably be the main challenge we might face. 

I started this component refinement and already found 2 challenges:
1. Tooltip can receive props which will be passed to the Tooltip itself. Text inside tooltip is by default set as white. Tooltip popover and tooltip arrow down are by default black. Giving users the ability to render their own tooltip is basically saying that TruncatedText does not support a tooltip internally, you should create your own tooltip and pass it to us... But, this pretty much breaks the idea of building a component that if necessary puts a tooltip in the surroundings of text. Why? Well, this is very simple. How can we determine the Tooltip trigger? How can we determine the tooltip trigger text? How can we determine the transition effect of the tooltip and **most important:** how can we determine the classes that style the tooltip and truncate it? Giving users full ability of determining their tooltips is good for _composability_? Might be. What I am saying is that it is bad for our component work. Let's take this as example: TruncatedText can receive custom tooltip JSX.Element. Whenever textOverflows is true, we display the `renderTooltip` passed to this component. This would mean:

```
{textOverflows ? (
   renderTooltip()
) : (
   renderText()
)}
``` 

So far so good. Let's now, compare the code share with the code present in the file tagged with this thread.

- The current `TruncatedText` uses renderText() to render the `Tooltip` trigger, including, then, `[textStylingClasses, textDefaultClasses]` in the classes of the informed `text` (as well as `maxDisplayWidth`)
- The shared code would leave up to people to render and customize the specs of the `Tooltip` of theirs. How can we guarantee they will truncate the trigger element? How can we force `maxDisplayWidth` of their trigger element? How can we guarantee all the specs that makes our TruncatedText, more than a simple paragraph element? --- We can surely guarantee we will render their JSX.Element based on their `renderTooltip` function. We can surely guarantee whatever they pass to us as `thisProps` or `thereProps` will be applied `this` and `there`. We cannot, though, guarantee the trigger element of their tooltip will be truncated as we want since whenever we say "Tooltip" we say "attached to", which means: how can we ensure integrators tooltip will be attached to our trigger element? The only option I see is by creating A LOT of complexity around how to specify `TruncatedText`. The idea of creating one component is to make it easy for integrators to reuse it and, after doing local tests, I share my opinion that we might be wanting to go to far with composability in this component. IMO, what would make an ideal TruncatedText with the thinest layer possible would be to remove our Tooltip component from it and add a `onTextOverflows` callback which is triggered whenever the informed text, with the informed CSS classes, with the informed maxDisplayWidth: overflows. This way we can create a component that reuses TruncatedText and our Tooltip (setting the tooltip whenever `onTextOverflows` is called) and leave up to the users the ability of building the Tooltip as they want to.

2. The problem I see is that we do not want our tooltip declared internally in the component! The goal I see is that we want integrators to have full control over the Tooltip to be used! The resolution I see is shared above.

Reply:
Hi!

Happy to share I have created this story to keep the progress on this thread's subject:

[TruncatedText with custom integrators Tooltip?](https://app.shortcut.com/ps-web3/story/25710/truncatedtext-with-custom-integrators-tooltip)